### DISCUSSION TITLE
Problems with runtime logger in parallel tasks

### DISCUSSION THREAD

**Discussion by Nathan-Intergral** [microsoft/autogen#2617] [General]
Hey all,

I am having some trouble with the runtime logger not working as expected when running several of them in parallel in my application.

The problem I am seeing is that all of the .db files are created correctly to start with but only one of them ends up with the chat completions table populated at the end.

I noticed that the `runtime_logging.py` file initialises the logger in to a global variable as shown below:

```logger = logging.getLogger(__name__)

autogen_logger = None
is_logging = False


def start(logger_type: str = "sqlite", config: Optional[Dict[str, Any]] = None) -> str:
    global autogen_logger
    global is_logging

    autogen_logger = LoggerFactory.get_logger(logger_type=logger_type, config=config)

    try:
        session_id = autogen_logger.start()
        is_logging = True
    except Exception as e:
        logger.error(f"[runtime logging] Failed to start logging: {e}")
    finally:
        return session_id
```

I believe this may be the issue? This means that only a single instance of the logger will be callable at one time as the autogen_logger variable will be overwritten each time the logger is initialised.

Thanks in advance for your cooperation.

**Comment by ekzhu** [comment]
You are correct there is only one logger. We are switching to different initialization method so the same logger will be used across all threads rather than creating a new logger in the `start` function. #2596. Would you like to take a look at the PR?

**Reply by Nathan-Intergral** [reply]
great thank you for the info! Looking at #2596 , will there be a way to differentiate between the different threads chat completions? Id need to be able to query the sqlite db for all chat completions of a single thread

**Reply by ekzhu** [reply]
I think you can use session ID returned by `start` to identify the session?

**Reply by Nathan-Intergral** [reply]
ah yes apologies, i would init the logger before executing the parallel processes and call start in each threads process individually giving me a different session id, is that correct?

**Reply by ekzhu** [reply]
For now, you want to create a logger in each thread as the `session_id` is created in the constructor of the logger.

**Reply by ekzhu** [reply]
Scratch that... My previous comment was wrong. #2596  gives you the ability to use the same logger instance across different threads, but it gives you the same `session_id` across different threads. So you can't use `session_id` to identify your thread. To differentiate between different threads in your logs, we need to also log the thread ID. It is not currently logged. We should probably add it.

**Reply by ekzhu** [reply]
I created an issue: https://github.com/microsoft/autogen/issues/2656
